### DISCUSSION TITLE
useTokenPagination

### DISCUSSION THREAD

**Discussion by ggascoigne** [TanStack/table#1762] [General]
Looks like useTokenPagination hasn't been updated to match the new interface.

**Comment by tannerlinsley** [comment]
Indeed. It's also undocumented and pretty experimental. I've been visiting it with new ideas here and there, but I don't think it's ready for prime time yet.

**Comment by ggascoigne** [comment]
I had to fork an early version to get it doing what I needed - but the token style iteration seems very backend specific.

**Comment by tannerlinsley** [comment]
I would love any feedback or ideas you have for it. Yes I agree, it's pretty specific to something like apollo and graphql vs general token pagination.

**Comment by IanVS** [comment]
I'd really love an example of how to perform token pagination with something like graphql.  Is `useTokenPagination` intended to be used in conjunction with the standard `usePagination`?  It seems like it just exposes some utility helpers, and doesn't really hook into `react-table` at all...

**Comment by tannerlinsley** [comment]
`useTokenPagination` never really made the cut to v7 because of these issues. If it's still there, then we should remove it and put it in the wiki or something. It's really just a normal hook that can help you transpose cursor based pagination into an index-based one.

**Reply by IanVS** [reply]
OK, that's what it seemed like, but the docs for `usePagination` say to use it _instead_ if you're doing cursor pagination.  I've given it some more thought and I think all I really need is a "load more" button below my table, so I think I can skip true pagination entirely.

**Comment by egemon** [comment]
Any updates on these guys? Any workaround etc?

**Comment by TastyDucks** [comment]
@egemon @sparkle32817
If your pagination token cursor is a unique value present in the returned data (like a UUIID), once you set `manualPagination = true`, you can probably calculate all of your [pagination values](https://react-table.tanstack.com/docs/api/usePagination#instance-properties) inside react-table via:

- Total number of available records on the server
- `rowIndex` of the record described by the pagination token
- Number of records per page

If your client can't parse the cursor, you will have to rely on other information to manage pagination -- any sort of count details the server returns, whether it returns an empty array or `null` as your client requests more details, etc. Depends on implementation.

Generally, it seems to me that the implementation details of remote pagination is outside the scope of react-table, and [the reference](https://github.com/tannerlinsley/react-table/blob/8c77b4ad97353a0b1f0746be5b919868862a9dcc/docs/src/pages/docs/api/usePagination.md#L8) to `useTokenPagination` in the docs is probably just left-over from a previous version.

Finally, I may be implementing something like this myself shortly; if I get permission, I'll share an example in this thread, or maybe as a react-query plugin if its polished enough.

**Reply by maxgurewitz** [reply]
> rowIndex of the record described by the pagination token

Cursor based pagination is typically implemented where an offset / `rowIndex` can't be calculated for a particular cursor token.

**Comment by siegenthalerroger** [comment]
Just wondering if there's been any updates on TanStack Table's handling of token-based pagination? Considering it's only really relevant in cases where the "total numer of available records" isn't known, @TastyDucks' answer isn't really helpful. I understand if it's still considered out of scope, just wondering if there's an update.

**Comment by lukeramsden** [comment]
This makes tanstack/table pretty useless if your API uses cursor-based pagination (which is a LOT of APIs!, Stripe as an example https://docs.stripe.com/api/pagination). How am I supposed to paginate in a cursor-based manner rather than an offset-based manner?

**Comment by maxgurewitz** [comment]
In fact, cursor based pagination is not only common, but it's practically required for larger datasets in partitioned data stores. Any guidance here would be incredibly helpful!

**Comment by lukeramsden** [comment]
I ended up controlling pagination and filtering manually completely outside of tanstack/table, and wrote a generic hook for paginating over my Stripe-style cursor based list endpoints https://gist.github.com/lukeramsden/003cf887ab29a7faeb09d8f0c23a4b19
